### PR TITLE
Refactor PoC for chat client and server

### PR DESCRIPTION
**Pull request by TobeTek**

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v2.5.0
```
Comment (ooliver1) [inline .pre-commit-config.yaml]: This should probably be in its own PR.

File: poc/events.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+class Event:
+    """
+    Base class for all events
+
+        - Implements some useful helper functions
+        - Useful for exception handling/type hints
+    """
+
+    @classmethod
+    def from_dict(cls, data: Dict[Any, Any]):
```
Comment (ooliver1) [inline poc/events.py]:
```suggestion
    def from_dict(cls, data: Dict[str, Any]):
```
field names are always string

File: poc/ws_client.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            async_loop,
+            client_connection.handler,
+        ),
+        daemon=True,
+    ).start()
+    client_gui.start_mainloop()
+
+
+"""
+SIMPLE CLIENT
```
Comment (ooliver1) [inline poc/ws_client.py]: Why are your single-line docstrings on 3 lines, the [PEP 257](https://peps.python.org/pep-0257/) conventions say:

```py
"""single line summary"""
```

```py
"""single line summary

optional larger
description
"""
```

File: poc/ws_client.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        """
+        # Do something with the received msg
+        print(msg, end="\n>>>")
+
+
+class ClientConnection:
+    """Websocket Client Connection"""
+
+    def __init__(
+        self, ws_connection, producer: Awaitable, consumer: Awaitable, *args, **kwargs
```
Comment (ooliver1) [inline poc/ws_client.py]: Typehint `ws_connection` and use a generic on `Awaitable` to denote the return types. Also why is `*args, **kwargs` added when never used?

**Review by ooliver1 — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (TobeTek) [reply .pre-commit-config.yaml]: Came up as I was fixing some pre-commit errors. 
I'll checkout to the last version of the `.pre-commit-config.yaml` and we should be good?

File: poc/ws_client.py (hunk 3 above)
Comment (TobeTek) [reply poc/ws_client.py]: They're serving as headers/comments here, not necessarily docstrings.

File: poc/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+`events.py` : Sample Event base class and Event types
+
+`server.py` : WebSockets Server implemented with FastAPI
+
+`gui.py` : Tkinter Chat Client GUI
+
+`ws_client.py` : An implementation of websocket client with *`websockets`* library. Supports calls to a blocking "producer".
+
+# To Run
```
Comment (ooliver1) [inline poc/README.md]: markdown should not include more than one level 1 header
```suggestion
## To Run
```

File: poc/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+Start Client
+```bash
+> poetry run python poc/ws_client.py
+```
+
+Uncomment as appropriate to run the CLI or GUI chat client
+
+# TODO
+
+ - Ready to go 🚀
```
Comment (ooliver1) [inline poc/README.md]:
```suggestion
Uncomment as appropriate to run the CLI or GUI chat client

```

File: poc/events.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        Args:
+            data (Dict[Any, Any]): Fields and values for event type
+
+        Returns:
+            Event: an instance of an event
+        """
+        return cls(**data)
+
+    def to_dict(self):
```
Comment (ooliver1) [inline poc/events.py]: Include an explicit return type
```suggestion
    def to_dict(self) -> Dict[str, Any]:
```

File: poc/ws_client.py (hunk 3 above)
Comment (ooliver1) [reply poc/ws_client.py]: But still, why do they need 3 entire lines for a one line comment?

**Review by ooliver1 — CHANGES_REQUESTED** [review]

**Comment by c-gaskell** [comment]
I think it would be much cleaner to have separate submodules for client and server... ie their own folders with __init__.py etc. This means if multiple files are needed they're kept separate

**Comment by TobeTek** [comment]
> 

The final implementation would have a far better code structure. There are already existing folders/modules for `client` and `server`.

**Review by ooliver1 — APPROVED** [review]
Good enough for me as a first implementation

**Pull request merged**